src/ocean/math/WideUInt.d

```diff
@@ -0,0 +1,591 @@
+
+        Enables equality comparison with WideUInt of the same size
+
+        Params:
+            rhs = value to compare against
+
+    ***************************************************************************/
+
+    public equals_t opEquals ( WideUInt rhs )
+    {
+        return this.payload[] == rhs.payload[];
+    }
+
+    unittest
+    {
+        WideUInt a, b;
+        test!("==")(a, b);
+        foreach (size_t i, ref elem; a.payload)
+        {
+            a.payload[i] = 1337;
+            b.payload[i] = 1337;
+        }
+        test!("==")(a, b);
+        a.payload[$-1] = 0;
+        test!("!=")(a, b);
+    }
+
+    /**************************************************************************/
+
+    mixin(genOpCmp("
+    {
+        ptrdiff_t idx = N-1;
+        while (idx > 0)
+        {
+            if (this.payload[idx] != 0 || rhs.payload[idx] != 0)
+                break;
+            --idx;
+        }
+
+        auto a = this.payload[idx];
+        auto b = rhs.payload[idx];
+        return a < b ? -1 : (a > b ? 1 : 0);
+    }
+    "));
+
+    unittest
+    {
+        WideUInt num1, num2;
+        num1.payload[1] = 1;
+        num2.payload[2] = 1;
+        test(num1 < num2);
+        test(num2 > num1);
+    }
+
+    /***************************************************************************
+
+        Increment current value by one fitting in uint
+
+        Params:
+            rhs = value to increment with. Limited to `uint` for now to simplify
+                internal arithmetic.
+
+        Throws:
+            WideUIntRangeException if this WideUInt value range was to be
+            exceeded
+
+    ***************************************************************************/
+
+    public void opAddAssign ( uint rhs )
+    {
+        if (add(this.payload[0], rhs))
+            enforce(.wideint_exception, this.checkAndInc(1));
+    }
+
+    unittest
+    {
+        WideUInt num = 0;
+        for (auto i = 0; i < 1000; ++i)
+            num += (1 << 31);
+        test!("==")(num.payload[0], 0);
+        test!("==")(num.payload[1], 500);
+    }
+
+    /***************************************************************************
+
+        Mutates current number in-place, dividing it by `rhs` and calculating
+        remainder.
+
+        Params:
+            rhs = number to divide by
+
+        Returns:
+            remainder of division
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
+    {
+        ulong remainder = 0;
+
+        for (ptrdiff_t idx = this.payload.length - 1; idx >= 0; --idx)
+        {
+            remainder = (remainder << 32) + this.payload[idx];
+            ulong result = remainder / rhs;
+            remainder -= rhs * result;
+            assert(result <= uint.max);
+            this.payload[idx] = cast(uint) result;
+        }
+
+        assert(remainder <= uint.max);
+        return cast(uint) remainder;
+    }
+
```

Comment:
Omitted intentionally for now. I would prefer to add "meat to bones" (various opeator overloads) separately from core impl.